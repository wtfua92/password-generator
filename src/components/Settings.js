import React from 'react';
import PropTypes from 'prop-types';
import {types} from '../tools/constants';

Settings.propTypes = {
    settings: PropTypes.shape({
        passwordLength: PropTypes.number,
        number: PropTypes.bool,
        upperCase: PropTypes.bool,
        symbol: PropTypes.bool,
    }),
    changeSettingsHandler: PropTypes.func,
    changePasswordLengthHandler: PropTypes.func
};

function Settings({
    settings: {
        passwordLength,
        number,
        upperCase,
        symbol,
    },
    changeSettingsHandler,
    changePasswordLengthHandler
}) {
    return (
        <div className="password-generator__settings">
            <div>
                <label htmlFor="passwordLength">Length: </label>
                <input
                    id='passwordLength'
                    type="number"
                    value={passwordLength}
                    onChange={({target}) => {changePasswordLengthHandler(parseInt(target.value));}}
                />
            </div>
            <div>
                <label htmlFor="useNumber">
                    <input id="useNumber" type="checkbox" value={number} onChange={() => { changeSettingsHandler(types.NUMBER.toLowerCase()); }}/>
                    use number
                </label>
            </div>
            <div>
                <label htmlFor="useUpperCase">
                    <input id="useUpperCase" type="checkbox" value={upperCase} onChange={() => { changeSettingsHandler(types.UPPERCASE.toLowerCase()); } }/>
                    use uppercase
                </label>
            </div>
            <div>
                <label htmlFor="useSymbol">
                    <input id="useSymbol" type="checkbox" value={symbol} onChange={() => { changeSettingsHandler(types.SYMBOL.toLowerCase()); }}/>
                    use special symbols
                </label>
            </div>
        </div>
    );
}

export default Settings;