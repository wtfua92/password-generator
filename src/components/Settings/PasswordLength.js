import React from "react";
import PropTypes from "prop-types";

function PasswordLength({ passwordLength, changePasswordLengthHandler }) {
  return (
    <div className="password-generator__settings__length">
      <label htmlFor="passwordLengthNumber">Length: </label>
      <input
        id="passwordLength"
        type="range"
        min={8}
        max={16}
        value={passwordLength}
        onChange={({ target }) =>
          changePasswordLengthHandler(parseInt(target.value))
        }
      />
      <input
        id="passwordLengthNumber"
        type="number"
        min={8}
        max={16}
        value={passwordLength}
        onChange={({ target }) =>
          changePasswordLengthHandler(parseInt(target.value))
        }
      />
    </div>
  );
}

PasswordLength.propTypes = {
  passwordLength: PropTypes.number.isRequired,
  changePasswordLengthHandler: PropTypes.func.isRequired
};

PasswordLength.defaultProps = {
  passwordLength: 8,
  changePasswordLengthHandler: () => {}
};

export default PasswordLength;
