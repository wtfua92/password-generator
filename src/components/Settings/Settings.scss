@import "../../assets/styles/colors";

.password-generator {
  &__settings {
    margin-bottom: .75rem;

    &__item {
      label {
        display: flex;
        align-items: center;
        margin-bottom: 0.25rem;
        cursor: pointer;
      }

      input[type="checkbox"] {
        -webkit-appearance: none;
        -moz-appearance: none;
        appearance: none;
        width: 1rem;
        height: 1rem;
        background-color: $white;
        outline: none;
        cursor: pointer;
        transition-duration: 0.3s;

        &:checked {
          position: relative;
          background-color: $success;
        }
      }

      span {
        padding-left: 0.25rem;
      }
    }

    &__length {
      margin-bottom: 0.75rem;
      display: flex;
      justify-content: space-between;
      align-items: center;
      input[type=range] {
        -webkit-appearance: none;
        width: 60%;
        margin: 13.8px 0.5rem;
      }
      input[type=range]:focus {
        outline: none;
      }
      input[type=range]::-webkit-slider-runnable-track {
        width: 100%;
        height: 8.4px;
        cursor: pointer;
        background: $success;
        border-radius: 1.3px;
        border: 0.2px solid $black;
      }
      input[type=range]::-webkit-slider-thumb {
        height: 24px;
        width: 24px;
        border-radius: 50%;
        background: #ffffff;
        cursor: pointer;
        -webkit-appearance: none;
        margin-top: -8px;
      }
      input[type=range]:focus::-webkit-slider-runnable-track {
        background: $success;
      }
      input[type=range]::-moz-range-track {
        width: 100%;
        height: 8.4px;
        cursor: pointer;
        box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
        background: $success;
        border-radius: 1.3px;
        border: 0.2px solid #010101;
      }
      input[type=range]::-moz-range-thumb {
        box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
        border: 1px solid #000000;
        height: 24px;
        width: 24px;
        border-radius: 50%;
        background: #ffffff;
        cursor: pointer;
      }
      input[type=range]::-ms-track {
        width: 100%;
        height: 8.4px;
        cursor: pointer;
        background: transparent;
        border-color: transparent;
        color: transparent;
      }
      input[type=range]::-ms-fill-lower {
        background: $success;
        border: 0.2px solid #010101;
        border-radius: 2.6px;
        box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
      }
      input[type=range]::-ms-fill-upper {
        background: $success;
        border: 0.2px solid #010101;
        border-radius: 2.6px;
        box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
      }
      input[type=range]::-ms-thumb {
        box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
        border: 1px solid #000000;
        width: 16px;
        border-radius: 3px;
        background: #ffffff;
        cursor: pointer;
        height: 8.4px;
      }
      input[type=range]:focus::-ms-fill-lower {
        background: $success;
      }
      input[type=range]:focus::-ms-fill-upper {
        background: $success;
      }

      input[type="number"] {
        width: 20%;
        margin-left: .5rem;
        text-align: center;
        outline: none;
        background-color: transparent;
        border: none;
        border-bottom: 1px solid $white;
        color: $white;
        &:active, &:focus {
          border-bottom: 1px solid $success;
        }
      }
    }
  }
}