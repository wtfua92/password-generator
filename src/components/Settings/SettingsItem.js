import React from "react";
import PropTypes from "prop-types";

function SettingsItem({ value, onChangeHandler, type }) {
  return (
    <div className="password-generator__settings__item">
      <label htmlFor={`setting-item-${type}`}>
        <input
          id={`setting-item-${type}`}
          type="checkbox"
          value={value}
          onChange={() => onChangeHandler(type)}
        />
        <span>use {type}</span>
      </label>
    </div>
  );
}

SettingsItem.defaultProps = {
  value: false,
  onChangeHandler: () => {},
  type: ""
};

SettingsItem.propTypes = {
  value: PropTypes.bool,
  onChangeHandler: PropTypes.func,
  type: PropTypes.string
};

export default SettingsItem;
