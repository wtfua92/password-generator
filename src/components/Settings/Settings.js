import React from "react";
import PropTypes from "prop-types";

import "./Settings.scss";

import SettingsItem from "./SettingsItem";
import PasswordLength from "./PasswordLength";

function Settings({
  settings,
  passwordLength,
  changeSettingsHandler,
  changePasswordLengthHandler
}) {
  return (
    <div className="password-generator__settings">
      <PasswordLength
        passwordLength={passwordLength}
        changePasswordLengthHandler={changePasswordLengthHandler}
      />
      {Object.keys(settings).map(type => (
        <SettingsItem
          value={settings[type]}
          key={type}
          onChangeHandler={() => changeSettingsHandler(type)}
          type={type}
        />
      ))}
    </div>
  );
}

Settings.propTypes = {
  passwordLength: PropTypes.number.isRequired,
  settings: PropTypes.object.isRequired,
  changeSettingsHandler: PropTypes.func.isRequired,
  changePasswordLengthHandler: PropTypes.func.isRequired
};

Settings.defaultProps = {
  settings: {},
  passwordLength: 8,
  changeSettingsHandler: () => {},
  changePasswordLengthHandler: () => {}
};

export default Settings;
