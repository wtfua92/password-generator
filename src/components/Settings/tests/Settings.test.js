import React from "react";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import Settings from "../Settings";

describe("<Settings />", () => {
  const changeSettingsHandlerMock = jest.fn();
  const changePasswordLengthHandlerMock = jest.fn();
  const props = {
    settings: {
      integer: false,
      symbol: false,
      uppercase: false
    },
    passwordLength: 12,
    changeSettingsHandlerMock: changeSettingsHandlerMock,
    changePasswordLengthHandlerMock: changePasswordLengthHandlerMock
  };
  let container;

  beforeEach(() => {
    container = render(<Settings {...props} />);
  });

  afterEach(cleanup);

  it("should include password length section", () => {
    const { getByLabelText } = container;
    const passwordLength = getByLabelText("Length:");
    expect(passwordLength).toBeDefined();
  });

  it("should have as many setting items as there are key:value pairs in settings prop", () => {
    const { getAllByLabelText } = container;
    const settingItems = getAllByLabelText(/^use/);
    expect(settingItems.length).toEqual(Object.keys(props.settings).length);
  });

  it("should match snapshot", () => {
    const {
      container: { firstChild: root }
    } = container;
    expect(root).toMatchSnapshot();
  });
});
