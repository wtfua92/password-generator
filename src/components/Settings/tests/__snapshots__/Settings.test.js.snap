// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> should match snapshot 1`] = `
<div
  class="password-generator__settings"
>
  <div
    class="password-generator__settings__length"
  >
    <label
      for="passwordLengthNumber"
    >
      Length: 
    </label>
    <input
      id="passwordLength"
      max="16"
      min="8"
      type="range"
      value="12"
    />
    <input
      id="passwordLengthNumber"
      max="16"
      min="8"
      type="number"
      value="12"
    />
  </div>
  <div
    class="password-generator__settings__item"
  >
    <label
      for="setting-item-integer"
    >
      <input
        id="setting-item-integer"
        type="checkbox"
        value="false"
      />
      <span>
        use 
        integer
      </span>
    </label>
  </div>
  <div
    class="password-generator__settings__item"
  >
    <label
      for="setting-item-symbol"
    >
      <input
        id="setting-item-symbol"
        type="checkbox"
        value="false"
      />
      <span>
        use 
        symbol
      </span>
    </label>
  </div>
  <div
    class="password-generator__settings__item"
  >
    <label
      for="setting-item-uppercase"
    >
      <input
        id="setting-item-uppercase"
        type="checkbox"
        value="false"
      />
      <span>
        use 
        uppercase
      </span>
    </label>
  </div>
</div>
`;
